### PR TITLE
feat: implement "relations" feature to MongoDB (just like "populate" in Mongoose)

### LINKED ISSUE #655: Referenced object doesn't get initialized (MongoDB)
Hello everyone,

I am currently implementing a basic user/group system with TypeORM on MongoDB. I got the following relations:

1. Many users got one group
2. One group got many users

Everything gets saved correctly, the problem is that the referenced group object in the user object doesn't get loaded when I am loading my user object from the database.

**Small example:**
user.entity.ts
```ts
@Entity()
export class User {
    @ObjectIdColumn()
    id: string;

    @Column()
    name: string;

    @Column("string")
    @Index({unique: true})
    email: string;

    @Column()
    password: string;

    @ManyToOne(type => Group, group => group.users)
    group: Group;

    @Column()
    code_activation: string;

    @Column()
    code_forgotmail: string;

    @CreateDateColumn()
    registered: string;
}
```

group.entity.ts
```ts
@Entity()
export class Group {
    @ObjectIdColumn()
    id: string;

    @Column("int")
    power: number;

    @Column("string")
    @Index({unique: true})
    name: string;

    @OneToMany(type => User, user => user.group)
    users: User[];

    @UpdateDateColumn()
    updated: string;
}
```

testme.ts
```ts
let repository = await this.db.getRepository(User);
let user = await repository.findOne({email: email});

return user.group // is null
```

Best Regards

### PR DESCRIPTION
**Pull request by mrgoonie**
- refactor `MongoEntityManager`, use `aggregate` instead of `find*` to utilize `$lookup` for populating columns
- add test: `functional/mongodb/basic/relations` to test mongodb querying with "relations"

Closes: #655

### Description of change

Implement a "relations" feature of MongoDB query (just like "populate" in Mongoose)
It should close the issue #655 after 5 years of waiting :)
Add some examples to the documentation of MongoDB.

#### Using "relations" to populate referenced columns

`ManyToOne/OneToMany/ManyToMany/OneToOne` are only supported in relational databases. Mongodb is relational-less, using "relations" is some kind of "anti-pattern", but it could be useful in some case. If you really want to use this feature, please check the example below:

Define your entity as following:

```typescript
@Entity()
export class User {
    @ObjectIdColumn()
    _id: ObjectID

    @Column()
    name: string
}

@Entity()
export class Post {
    @ObjectIdColumn()
    _id: ObjectID

    @Column()
    title: string

    /**
     ** @ObjectIdColumn({ name: "user" })
     ** This means it's an referenced ObjectId of a {User} in "user" collection.
     */
    @ObjectIdColumn({ name: "user" })
    author: ObjectID | User
}
```

Create a user & a post:

```typescript
import { getMongoManager } from "typeorm"

const manager = getMongoManager()

const user = new User()
user.name = "Justin"
await manager.save(user)

const post = new Post()
post.author = user._id
post.title = "About the sky"
await manager.save(post)
```

Then you can populate `author` column with `relations` options (like ORM):

```typescript
const posts = await myDataSource.getMongoRepository(Post).find({
    relations: {
        author: true,
    },
})
```

If you want to use `ManyToOne/OneToMany/ManyToMany`, take this example below for defining an entity:

```typescript
@Entity()
export class User {
    @ObjectIdColumn()
    _id: ObjectID

    @Column()
    name: string

    // Make sure you have {array: true}
    @ObjectIdColumn({ name: "post", array: true })
    posts: ObjectID[] | Post[]
}

// create new {User} with something like this:
const user = new User()
user.name = "Selena"
user.posts = [
    new ObjectID("89903245..."),
    new ObjectID("11111111..."),
    new ObjectID("22222222..."),
]
```

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run format` to apply prettier formatting
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #655`
- [x] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by hghammoud** [comment]
can we please accept this PR? we all been waiting for so long for this.

**Comment by hghammoud** [comment]
@mrgoonie I check out your branch and built it. I used yalc to add it locally. Can you please give a more concrete example of nested lists as the relations is not populating the object

```
import { Page } from 'src/page/entities/page.entity';
import { Column, Entity, ObjectID, ObjectIdColumn } from 'typeorm';

@Entity()
export class Namespace {
    @ObjectIdColumn()
    _id: ObjectID;

    @Column()
    label: string = 'Untitled';

    @Column()
    icon: string;

    @ObjectIdColumn({ name: "pages", array: true })
    pages: ObjectID[] | Page[];
}
```
```
import { Namespace } from 'src/namespaces/entities/namespace.entity';
import { Column, Entity, ObjectID, ObjectIdColumn } from 'typeorm';

@Entity()
export class Page {
    @ObjectIdColumn()
    _id: ObjectID;

    @Column()
    title: string = 'Undefined';

    @Column()
    isGroup?: boolean = false;

    @Column()
    markdown?: string = '';

    @ObjectIdColumn({ name: "namespace" })
    namespace?: ObjectID | Namespace;
}
```
```
  @Post()
  async create(@Body() createNamespaceDto: CreateNamespaceDto) {
    let page = new Page();
    page = await this.pageService.create(page);

    let nameSpace: Namespace = CreateNamespaceDto.toNamespace(createNamespaceDto);
    nameSpace.pages = [
      page._id
    ];
    nameSpace = await this.namespacesService.create(nameSpace);

    return await this.namespacesService.findOne(nameSpace._id.toString());
  }
```
```
  findOne(id: string) {
    return this.namespaceRepository.findOne({
      relations: {
        pages: true,
      },
      where: { _id: new ObjectId(id) } as any
    });
  }
```

the result of the the findOne is an empty array... what am I doing wrong? ty.

**Comment by mrgoonie** [comment]
hey @hghammoud, in "Namespace" entity, change your column decorator "name" to "page" instead of "pages", since it will populate from "Page" collection.

Like this:
```
@ObjectIdColumn({ name: "page", array: true })
pages: ObjectID[] | Page[];
```

then it should work as expected

**Comment by hghammoud** [comment]
I didnt test further then this case but that did the trick. I ll tell you if other issues arise. ty.

**Comment by hghammoud** [comment]
It worked. I will be testing further and give you feedback if something is off.

File: docs/mongodb.md
Comment (gastonmorixe) [inline docs/mongodb.md:187]:
```suggestion
     ** This means it's a referenced ObjectId of a {User} in "user" collection.
```

File: test/functional/mongodb/basic/array-columns/mongodb-array-columns.ts
Comment (gastonmorixe) [inline test/functional/mongodb/basic/array-columns/mongodb-array-columns.ts:95]:
```suggestion
```

**Review by gastonmorixe — CHANGES_REQUESTED** [review]

**Comment by pleerock** [comment]
I need mongodb users review & approval on this PR.

**Comment by mrgoonie** [comment]
I've made another pull request to fix some issues: https://github.com/typeorm/typeorm/pull/9767

Cheers,

**Comment by dmitrye** [comment]
What's the latest here? 

I'm using an aggregateEntity workaround for the time being to populate the child object. 
```

export class UserLogin extends CommonEntityFields {
    ....

@Column((type) => User)
    user: User;

 ....
 }
 

``` 
And the service:

```
this.userLoginRepository.aggregateEntity([
            {
                $lookup: {
                    from: 'user',
                    localField: 'user',
                    foreignField: '_id',
                    as: 'user',
                }
            },
            {
                "$unwind": {
                    "path": "$user",
                    "preserveNullAndEmptyArrays": true
                }
            },
            {
                $match: filter
            }
        ]).toArray();
```

But this will become cumbersome to use Aggregate queries everywhere just to merge the relationship data. 
 
 TL;DR
The better Mongo approach is to avoid $lookups in favor of data duplication whenever possible. Things that can be considered snapshots in time: Events, reports, etc, should self contain everything they need from the time of the activity.

**Comment by pleerock** [comment]
I'm not sure if this PR is still valid, because @mptr just implemented mongodb v5 support and there are lot of conflicts right now. Please create a new PR on top of the latest changes and wait for the community review.

**Comment by dmitrye** [comment]
@pleerock Can you confirm that the PR for mongo v5 also addresses relationships? If not, then this issue should remain open and be dependent on https://github.com/typeorm/typeorm/pull/9778 being addressed first. 

The purpose of this ticket was to enable Mongo relationships for OneToOne, OneToMany, etc -> and general use case when you have Entity attributes that are lookups into other collections. Which is different from supporting v5 changes.

**Comment by pleerock** [comment]
no, I just said that this PR isn't valid because there were many conflicts with the new changes on this PR.

**Comment by angelhdzmultimedia** [comment]
Can someone provide info about if this is already possible?

```ts
@Entity({name: 'profiles'})
class Profile {
  ...someFields
}

@Entity({name: 'users'})
class User {
 ...someFields
 @ObjectIdColumn({name: 'profiles'})
 profile: Profile | ObjectId
}
```

Usage:

```ts
const newUser = this._userRepository.create(registerData)
    const newProfile = this._profileRepository.create({
      displayName: registerData.displayName,
      role: 'user',      
    })
    const createdProfile = await this._profileRepository.save(newProfile, {reload: true})
    const createdUser = await this._userRepository.save(newUser, { reload: true })
    createdUser.profile = createdProfile
    return createdUser
```

And then:

```ts
const user: User | null = await this._userRepository.findOne({
      where: {
        email: loginData.email,
      },
      relations: {
        profile: true,
      }
    })
```

Tried with `relations: ['profile']`.

**Pull request closed without merging**
